### PR TITLE
feat(bigquery): Add pushdown_deny_usernames and pushdown_allow_usernames for server-side user filtering

### PR DESCRIPTION
**Pull request by dinesh-verma-datahub**
## Summary

Adds server-side user email filtering (pushdown) for BigQuery queries, allowing users to filter query data at the BigQuery level before it's returned to DataHub. This significantly improves performance for large datasets by reducing data transfer.

This feature mirrors the existing Snowflake `pushdown_deny_usernames` / `pushdown_allow_usernames` implementation for cross-platform consistency.

**Related PR:** [connector-tests#612](https://github.com/acryldata/connector-tests/pull/612) - Integration tests for BigQuery pushdown filters

### New Configuration Options

| Field | Type | Description |
|-------|------|-------------|
| `pushdown_deny_usernames` | `List[str]` | SQL LIKE patterns to exclude (e.g., `bot_%`, `%@%.iam.gserviceaccount.com`) |
| `pushdown_allow_usernames` | `List[str]` | SQL LIKE patterns to include (e.g., `%@company.com`) |

**Requires:** `use_queries_v2: true`

### Pattern Syntax

Uses standard SQL LIKE syntax:
- `%` - matches any sequence of characters
- `_` - matches any single character
- Matching is **case-insensitive**

### Behavior

- **Deny wins:** If a user matches both allow and deny patterns, they are excluded
- **Empty allow list:** All users allowed (except those in deny list)
- **Server-side filtering:** Patterns are pushed to BigQuery as `LOWER(user_email) LIKE 'pattern'` clauses

### Example Usage

```yaml
source:
  type: bigquery
  config:
    use_queries_v2: true
    
    # Filter out service accounts and bots
    pushdown_deny_usernames:
      - "%@%.iam.gserviceaccount.com"
      - "bot_%"
      - "%_bot@%"
    
    # Only include company emails
    pushdown_allow_usernames:
      - "%@company.com"
      - "%@subsidiary.com"
```

### Generated SQL

```sql
WHERE (LOWER(user_email) NOT LIKE '%@%.iam.gserviceaccount.com')
  AND (LOWER(user_email) NOT LIKE 'bot_%')
  AND (LOWER(user_email) NOT LIKE '%_bot@%')
  AND (LOWER(user_email) LIKE '%@company.com' OR LOWER(user_email) LIKE '%@subsidiary.com')
```

### Files Changed

- `bigquery_config.py` - New config fields with validators
- `queries_extractor.py` - SQL filter generation logic
- `bigquery.py` - Wire config to extractor
- `bigquery_pre.md` - User documentation
- `test_bigquery_queries_extractor.py` - 50 unit tests

### Security

- SQL injection prevention via quote escaping (`'` → `''`)
- Empty/whitespace patterns rejected by validator
- Patterns validated before SQL generation

### Testing

- [x] 50 unit tests covering core logic, edge cases, and security
- [x] Integration tests in separate `connector-tests` repository

### Checklist

- [x] Consistent with Snowflake's pushdown implementation
- [x] Case-insensitive matching (matches Snowflake's ILIKE behavior)
- [x] SQL injection protection
- [x] Requires `use_queries_v2: true` (validated)
- [x] Documentation with examples
- [x] Pre-commit hooks pass

### CODE REVIEW

File: metadata-ingestion/docs/sources/bigquery/bigquery_pre.md [hunk 1]
```diff
@@ -139,6 +140,42 @@ source:
...
+    user_email_pattern:
+      allow:
+        - "analyst_.*@example\\.com"
+      deny:
+        - "bot_.*"
+```
+
+**Behavior:**
+
+- When enabled: Filtering happens in BigQuery SQL using `REGEXP_CONTAINS()`
```
Comment (sgomezvillamor) [inline metadata-ingestion/docs/sources/bigquery/bigquery_pre.md]: For analogy:

```suggestion
- When enabled: Filtering happens server-side with BigQuery SQL using `REGEXP_CONTAINS()`
```

File: metadata-ingestion/docs/sources/bigquery/bigquery_pre.md [hunk 2]
```diff
@@ -139,6 +140,42 @@ source:
...
 
+##### User Email Filtering Pushdown (Performance Optimization)
+
+The `pushdown_user_filter` option pushes down `user_email_pattern` filtering directly to BigQuery's SQL query, reducing data transfer and improving performance for large query volumes.
+
+**When to Use:**
+
+- You have large query volumes (>10k queries in your time window)
+- You're using `user_email_pattern` to filter users
+- You want to reduce BigQuery data transfer costs
```
Comment (sgomezvillamor) [inline metadata-ingestion/docs/sources/bigquery/bigquery_pre.md:151]:
```suggestion
- You want to reduce BigQuery data transfer costs
- You want to reduce overall DataHub ingestion time
```

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_config.py [hunk 3]
```diff
@@ -482,6 +482,15 @@ def have_table_data_read_permission(self) -> bool:
         "See [this](https://cloud.google.com/bigquery/docs/information-schema-jobs#scope_and_syntax) for details.",
     )
 
+    pushdown_user_filter: bool = Field(
+        default=False,
+        description="If enabled, pushes down `usage.user_email_pattern` filtering to BigQuery's "
```
Comment (sgomezvillamor) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_config.py]: `pushdown_user_filter` should not depend on `usage.*` configs, as those are specific for _usage_ feature.

We have several options here:

1. Adding `pushdown_deny_usernames` and `pushdown_allow_usernames`, as Snowflake connector does.
2. Adding `pushdown_username_pattern`
3. Other?

While 1 may not be ideal, I'd like to keep consistent configs across sources.
No strong opinion on that though and open to other options; my main concern is with `pushdown_user_filter` depending on `usage.*` configs.

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_config.py [hunk 4]
```diff
@@ -482,6 +482,15 @@ def have_table_data_read_permission(self) -> bool:
...
+        "INFORMATION_SCHEMA.JOBS query using REGEXP_CONTAINS for improved performance. "
+        "When disabled (default), filtering is done client-side using Python regex. "
+        "Enable this for large query volumes to reduce data transfer from BigQuery. "
+        "Only applicable if `use_queries_v2` is enabled.",
```
Comment (sgomezvillamor) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_config.py:491]: >        "Only applicable if `use_queries_v2` is enabled.",

Additional to the doc notice, this could become a pydantic model validator that may raise a warning or error if config is not consistent.

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/queries_extractor.py [hunk 5]
```diff
@@ -463,6 +482,139 @@ def close(self) -> None:
...
+    Returns:
+        True if the patterns effectively allow all, False otherwise
+    """
+    if not allow_patterns:
+        return True
+    if len(allow_patterns) == 1:
+        pattern = allow_patterns[0]
+        # Common patterns that match everything
+        return pattern in [".*", ".+", "^.*$", "^.+$", "[\\s\\S]*", "[\\s\\S]+"]
+    return False
```
Comment (sgomezvillamor) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/queries_extractor.py]: this would be also correct, right?

```suggestion
    return all(pattern in [".*", ".+", "^.*$", "^.+$", "[\\s\\S]*", "[\\s\\S]+"] for pattern in allow_patterns)
```

File: metadata-ingestion/tests/unit/bigquery/test_bigquery_queries_extractor.py [hunk 6]
```diff
@@ -0,0 +1,691 @@
...
+        """Multiple allow patterns should be OR'd together."""
+        pattern = AllowDenyPattern(
+            allow=["analyst_.*@example\\.com", "admin_.*@example\\.com"],
+            deny=[],
+            ignoreCase=False,
+        )
+        result = _build_user_filter_from_pattern(pattern)
+        assert "REGEXP_CONTAINS(user_email, 'analyst_.*@example\\\\.com')" in result
+        assert "REGEXP_CONTAINS(user_email, 'admin_.*@example\\\\.com')" in result
+        assert " OR " in result
```
Comment (sgomezvillamor) [inline metadata-ingestion/tests/unit/bigquery/test_bigquery_queries_extractor.py]: any reason why not checking with equality?

```
        assert result == "REGEXP_CONTAINS(user_email, 'analyst_.*@example\\\\.com') OR REGEXP_CONTAINS(user_email, 'admin_.*@example\\\\.com')"
```

whenever possible I would use equality for the validation, to check what's in and what is not in the result

File: metadata-ingestion/tests/unit/bigquery/test_bigquery_queries_extractor.py [hunk 7]
```diff
@@ -0,0 +1,691 @@
...
+        pattern = AllowDenyPattern(allow=["^.*$"], deny=[], ignoreCase=False)
+        result = _build_user_filter_from_pattern(pattern)
+        assert result == "TRUE"
+
+    def test_anchored_dot_plus_treated_as_allow_all(self):
+        """Pattern ^.+$ should be treated as allow-all."""
+        pattern = AllowDenyPattern(allow=["^.+$"], deny=[], ignoreCase=False)
+        result = _build_user_filter_from_pattern(pattern)
+        assert result == "TRUE"
+
```
Comment (sgomezvillamor) [inline metadata-ingestion/tests/unit/bigquery/test_bigquery_queries_extractor.py:249]: very good coverage in TestBuildUserFilterFromPattern
I just found that equality validation would be more appropriate in many cases

File: metadata-ingestion/tests/unit/bigquery/test_bigquery_queries_extractor.py [hunk 8]
```diff
@@ -0,0 +1,691 @@
...
+        """Query should include TRUE as default user filter."""
+        query = _build_enriched_query_log_query(
+            project_id="test-project",
+            region="region-us",
+            start_time=datetime(2024, 1, 1, tzinfo=timezone.utc),
+            end_time=datetime(2024, 1, 2, tzinfo=timezone.utc),
+        )
+        # Default filter is TRUE, but it should be in the WHERE clause
+        assert "TRUE" in query
+        assert "WHERE" in query
```
Comment (sgomezvillamor) [inline metadata-ingestion/tests/unit/bigquery/test_bigquery_queries_extractor.py]: `"TRUE" in query` may be too generic (non precise)

**Review by sgomezvillamor — COMMENTED** [review]
I left some comments

**Review by sgomezvillamor — APPROVED** [review]
🎖️
